drivers/siemens-sitop.c

```diff
@@ -0,0 +1,314 @@
+				upsfd = 0;
+			}
+			upsfd = ser_open_nf(device_path);
+			ser_set_speed_nf(upsfd, device_path, B9600);
+			done = 1;
+		} else if (num_received == 0) {
+			/* no (more) new data */
+			done = 1;
+		} else {
+			rx_count += num_received;
+			
+			/* parse received input data: */
+			while (rx_count >= (5 + NR_DELIMITER_CHARS)) { /* all input data are strings of 5 characters plus 1 or more delimiter characters */
+				if (strncmp(rx_buffer, "BUFRD", 5) == 0) {
+					current_ups_status.battery_alarm = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "ALARM", 5) == 0) {
+					current_ups_status.battery_alarm = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "DC_OK", 5) == 0) {
+					current_ups_status.dc_input_low = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "DC_LO", 5) == 0) {
+					current_ups_status.dc_input_low = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "*****", 5) == 0) {
+					current_ups_status.on_battery = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "*BAT*", 5) == 0) {
+					current_ups_status.on_battery = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "BA>85", 5) == 0) {
+					current_ups_status.battery_above_85_percent = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "BA<85", 5) == 0) {
+					current_ups_status.battery_above_85_percent = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else {
+					/* nothing sensible found at the start of the rx_buffer. */
+					rm_buffer_head(1);
+				}
+			}
+		}
+	}
+	return new_data_received;
+}
+
+
+static int instcmd(const char *cmdname, const char *extra) {
+	/* See the remarks in upsdrv_shutdown() */
+	if (!strcasecmp(cmdname, "shutdown.return")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command R");
+		ser_send_pace(upsfd, 200000, "\n\nR\n\n");
+		ser_send_pace(upsfd, 200000, "R\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command S");
+		ser_send_pace(upsfd, 200000, "\n\nS\n\n");
+		ser_send_pace(upsfd, 200000, "S\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+
+	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+
+
+void upsdrv_initinfo(void) {
+	int max_attempts = 5;
+	int found = 0;
+	while (!found && max_attempts > 0) {
+		if (check_for_new_data())
+			found = 1;
+		else
+			sleep(1); /* Sleep a while, then try again */
+		max_attempts--;
+	}
+	if (!found)
+		fatalx(EXIT_FAILURE, "No data received from the UPS");
+	
+	dstate_setinfo("ups.mfr", "Siemens");
```

Comment:
I'd remove these redundant 'ups.*' items.

Reply:
I got those from the skeleton driver skel.c, so that one might needs a revision as well.